src/set_theory/game/impartial.lean

```diff
@@ -19,10 +20,11 @@ universe u
 
 namespace pgame
 
-local infix ` ≈ ` := equiv
 local infix ` ⧏ `:50 := lf
+local infix ` ≈ ` := equiv
+local infix ` ∥ `:50 := fuzzy
 
-/-- The definition for a impartial game, defined using Conway induction -/
+/-- The definition for a impartial game, defined using Conway induction. -/
```

Comment:
Side comment: I think impartial games should just be defined as `G = -G`. Any definition using ≈ will not play well with misere play, for example.

Reply:
This seems extremely limiting. `{* | nim 1}` is very much an impartial game even if it isn't equal to its negative.

Reply:
I still disagree. Equality is a very finicky notion on our current definition of pre-games, and I'd rather not rely on it unless absolutely necessary.

Sure, this wouldn't work under Misere game convention, but neither would many other notions on pre-games. I wouldn't do this change based on that alone.

Reply:
Yes I just want to raise this point, which is definitely outside the scope of this PR. Maybe we'd introduce a type alias when we get to misere convention, since the setoid instance in normal convention doesn't make sense there.

Reply:
Actually under the current definition, impartial games are not necessary all-small, even though under the usual definition they are, and in particular all nimbers are. Of course they're still strongly infinitesimal by Sprague-Grundy.

Also the condition G ≈ -G is not easy to verify; it is itself defined recursively, so the current definition features nested recursion (of course the same also occur in the definition of `numeric`). However if you use G = -G (without other conditions!) then it's a single recursion. This condition will be automatically satisfied if you construct the `pgame` from a single type of positions with two well-founded relations L=R. That's why I said "suffice for most purposes".

A definition closer to the usual definition would be: for all left options, there exists a right option that is identical to it, and vice versa, and moreover all options are also impartial. But "identical" is also defined recursively, so won't be easy to verify unless the left options are actually equal to right options, in which case this would only allow reindexing the top level options in a game with G = -G.

Reply:
Oh, if this definition doesn't satisfy the usual theorems, then there's a problem that needs to be fixed.

That said, if Sprague Grundy works for these "impartial" games, it should be stated in terms of them. The added generality can only be a good thing.

Perhaps we can have two definitions? One for impartial games that uses the identity relation (which I plan to add at some point), and one for "identical" games that's the current definition. If we still take the typeclass route, we can easily prove impartial games are "impartial", so we can still use Sprague Grundy in its current form.

Reply:
In fact different references disagree in their definitions of *all_-mall*.

+ ONAG [conway2001] p.101 defines *all small* to mean that every position is small, and shows in Theorem 57 that G is all small iff every stopping position of G is (equivalent to) zero, where a *stopping position* is a position equivalent to a number. This definition includes my { | nim 1} example.

+ Siegel's *Combinatorial game theory* p.405 gives the same definition, but in p.60 defines G to be *dicotic* if both players can move from every nonempty subposition of G (i.e. same as Schleicher-Stoll's definition for *all-small*).

+ The beginning of Chapter 8 (in Vol. 1) of *Winning Ways* defines *all small* to mean that both players have legal moves from every non-terminal position, agreeing with Schleicher-Stoll.

Therefore I suggest we have separate definitions `all_small` and `dicotic`, and similarly a separate definition `impartial` and rename the current "impartial", so that we can show `all_small` `pgame`s are `impartial`, and `dicotic` games are "impartial".

As usual, a certain Library that rhymes with "Nemesis" has the references

Reply:
Sorry, the other way! `dicotic` implies the actual `impartial` and `all_small` implies the current "impartial".

Reply:
I've opened #14654, can we move this conversation there?

Reply:
Sure; sorry for hijacking this PR, but note that I already approved it.

I don't think Github allows to move discussion threads like this, so I think corrections to existing discussions are appropriate here, just for context. Subsequent new discussions should surely be moved to #14654.